File: java/p2/src/p2/types/operator/ScanJoin.java
package p2.types.operator;

import p2.lang.plan.Predicate;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.exception.P2RuntimeException;
import p2.types.table.Table;

public class ScanJoin extends Join {
	
	private Table table;
	
	public ScanJoin(Predicate predicate) {
		super(predicate);
		this.table = Table.table(predicate.name());
	}
	
	@Override
	public String toString() {
		return "NEST LOOP JOIN: PREDICATE[" + this.predicate  + "]";
	}
	
	@Override
	public TupleSet evaluate(TupleSet tuples) throws P2RuntimeException {
		TupleSet result = new TupleSet();
		TupleSet innerTuples = this.table.tuples();
		for (Tuple outer : tuples) {
			for (Tuple inner : innerTuples) {
				inner.schema(this.predicate.schema());
				
				if (validate(outer, inner)) {
					Tuple join = outer.join(inner);
					if (join != null) {
						result.add(join);
					}
				}
			}
		}
		return result;
	}

}

File: lincoln/ruby/lib/types/operator/scan_join.rb
require 'lib/types/operator/join'
require 'lib/types/table/table'

class ScanJoin < Join 
	def initialize(predicate, input)
		super(predicate, input)
		@table = Table.find_table(predicate.name)
		if @table.nil?
      require 'ruby-debug'; debugger
		  raise "ScanJoin initialization failed: couldn't find table " + predicate.name.to_s 
	  end
	end
	
	def to_s
		return "NEST LOOP JOIN: PREDICATE[" + @predicate.to_s  + "]"
	end
	
	def evaluate(tuples) #TupleSet
		result = TupleSet.new("tmp", nil)
		innerTuples = @table.tuples
		tuples.each do |outer| 
		  innerTuples.each do |inner|
				inner.schema = @predicate.schema.clone
				
				if validate(outer, inner)
					join = outer.join(inner)
					if (!join.nil?) 
						result << join
					end
				end
			end
		end

		return result
	end

end
